src/python/pants/engine/internals/graph.py

```diff
@@ -78,6 +81,39 @@
 # -----------------------------------------------------------------------------------------------
 
 
+@rule
+async def generate_subtargets(
+    address: Address, glob_match_error_behavior: GlobMatchErrorBehavior,
+) -> Subtargets:
+    if not address.is_base_target:
+        raise ValueError(f"Cannot generate file Targets for a file Address: {address}")
+    wrapped_base_target = await Get(WrappedTarget, Address, address)
+    base_target = wrapped_base_target.target
+
+    if not base_target.has_field(Dependencies) or not base_target.has_field(Sources):
+        # If a target type does not support dependencies, we do not split it, as that would prevent
+        # the base target from depending on its splits.
+        return Subtargets(base_target, ())
+
+    # Create subtargets for matched sources.
+    sources_field_path_globs = base_target[Sources].path_globs(glob_match_error_behavior)
+    if sources_field_path_globs is None:
+        return Subtargets(base_target, ())
+
+    # Generate a subtarget per source.
+    snapshot = await Get(Snapshot, PathGlobs, sources_field_path_globs)
```

Comment:
As discussed over DM, it's generally fine to not use `HydratedSources`. And I know we can't use that due to rule graph issues.

It doesn't matter that the `for_sources_type` thing is used, nor that codegen isn't used. That's fine.

The risk is that we won't be calling all the validation logic, like asserting that `python_binary` has 0-1 file. We could end up generating subtargets that are invalid. Later in the Pants run, if `HydrateSources` is used, we would normally error, but I'm not sure we would anymore because the `python_binary` would have been split into multiple generated subtargets.

A simple fix would be to call `my_sources.validate_snapshot` on the result.

Reply:
Adding. But that method will also end up being called for all subtargets when we Hydrate their sources. Something to thing about as we figure out `Target` API adjustments.